### PR TITLE
Implement focus trap behavior and hook

### PR DESCRIPTION
**Pull request by T-Hugs**
This PR adds a "focus trap" behavior. For more details on its behavior and API, please see the [rendered documentation](https://primer-components-qufu15vhg-primer.vercel.app/components/focusTrap).

### Screenshots
Screenshot from storybook story:
image

### Merge checklist
- [x] Added/updated tests
- [x] Added/updated documentation
- [x] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

### CODE REVIEW

File: src/hooks/useFocusTrap.ts
Comment (T-Hugs) [inline src/hooks/useFocusTrap.ts:11]: Add JSDoc comment.

File: src/polyfills/eventListenerSignal.ts
Comment (T-Hugs) [inline src/polyfills/eventListenerSignal.ts:1]: Add a comment to this file explaining what it does and referencing the spec

File: docs/content/focusTrap.mdx
Comment (T-Hugs) [inline docs/content/focusTrap.mdx:46]: Set the "Default" cell to _See description_.

File: docs/content/focusTrap.mdx
Comment (emplums) [inline docs/content/focusTrap.mdx:56]: This reminds me, initially we were planning on baking this into `Overlay` by default but it should probably be optional since we shouldn't be trapping focus in things like ActionMenu or Dropdown

File: src/__tests__/behaviors/focusTrap.tsx
Comment (emplums) [inline src/__tests__/behaviors/focusTrap.tsx:15]: Curious why you needed to put `tabIndex={0}` on these, shouldn't they already by tubbable since they are buttons?

File: src/utils/iterateTabbable.ts
Comment (emplums) [inline src/utils/iterateTabbable.ts:61]: I wasn't sure so I doubled checked in dev tools, it looks like if you explicitly set a tab index of 0 on an element, then set the disabled attribute, the tab index still comes back as 0. 
image

In this case, we'd return true though, even if `disabled` is set. I don't think we want to return early here, instead we should probably move lines 66 to 77 above this check and return early if any of the not tabbable checks are true

File: docs/content/focusTrap.mdx
Comment (emplums) [inline docs/content/focusTrap.mdx:51]: Would it make sense to build something into the JS behavior to abort the focus trap when the container is hidden or removed from the DOM?

File: src/behaviors/focusTrap.ts
Comment (emplums) [inline src/behaviors/focusTrap.ts:68]: This shouldn't be a problem with the Hook since we will re-run the hook whenever `containerRef` changes, but do we need to worry about containers that have dynamic content in them? (DM me for an example in an unreleased GitHub project.)

File: src/behaviors/focusTrap.ts
Comment (emplums) [inline src/behaviors/focusTrap.ts:75]: Should we run this through `isTabbable` instead of only checking if the tab index isn't `-1`?

File: docs/content/focusTrap.mdx
Comment (T-Hugs) [reply docs/content/focusTrap.mdx:46]: No problem, this isn't really a full hook. Might be able to refactor this to use yours 👍

File: docs/content/focusTrap.mdx
Comment (T-Hugs) [reply docs/content/focusTrap.mdx:51]: I could set up a mutation observer on the parent to ensure clean-up when the container is removed. I would kind of prefer to leave that to consumers though. Maybe we can have a check in the `focusin` event handler that makes sure the container is still around and focusable, and if not clean ourselves up? I'm wondering if this would be unexpected behavior in some way.

File: docs/content/focusTrap.mdx
Comment (T-Hugs) [reply docs/content/focusTrap.mdx:56]: I don't think Overlay should trap focus - that should be left to Dialogs.

File: docs/content/focusTrap.mdx
Comment (T-Hugs) [reply docs/content/focusTrap.mdx:74]: Makes sense 👍

File: src/__tests__/behaviors/focusTrap.tsx
Comment (T-Hugs) [reply src/__tests__/behaviors/focusTrap.tsx:15]: Because Jest/JSDOM 😂

Comment (T-Hugs) [reply src/__tests__/behaviors/focusTrap.tsx:15]: Actually, I might be able to remove them. I just figured since it's unrelated to what I'm actually testing, I didn't want to leave it to chance. I think I saw something on Stackoverflow where someone said that they had to include `tabindex="0"` on everything they wanted to call `.focus()` on.

File: src/behaviors/focusTrap.ts
Comment (T-Hugs) [reply src/behaviors/focusTrap.ts:75]: Yeah, good call. I wrote this before I wrote `isTabbable` and forgot to update it!

File: src/utils/iterateTabbable.ts
Comment (T-Hugs) [reply src/utils/iterateTabbable.ts:61]: Oh wow, so `disabled` trumps `tabindex="0"`, eh? That seems bad. We should probably avoid and advise against using the `disabled` attribute on form elements, instead opting for `aria-disabled` and relevant styling/behavior. Will fix.

ARIA actually recommends allowing disabled buttons to be focusable in some situations, which is why I assumed `tabindex="0"` would win here. See: https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_disabled_controls

File: src/__tests__/behaviors/focusTrap.tsx
Comment (T-Hugs) [reply src/__tests__/behaviors/focusTrap.tsx:15]: Ah, nope, confirmed that Jest won't let you `.focus()` any element that doesn't explicitly have tabindex set to a non-negative value.

File: docs/content/focusTrap.mdx
Comment (emplums) [reply docs/content/focusTrap.mdx:70]: fair enough!

File: src/__tests__/behaviors/focusTrap.tsx
Comment (emplums) [reply src/__tests__/behaviors/focusTrap.tsx:15]: Ok sounds good!

File: docs/content/focusTrap.mdx
Comment (emplums) [reply docs/content/focusTrap.mdx:51]: I think it's fair to leave it for now, was just curious if you had considered it

File: src/behaviors/focusTrap.ts
Comment (emplums) [reply src/behaviors/focusTrap.ts:68]: If you decide to do this in a follow up can you create an issue for it just in case we forget? This behavior will need to be supported in Memex

File: src/utils/iterateTabbable.ts
Comment (emplums) [reply src/utils/iterateTabbable.ts:69]: From what I understand, the tab index attribute returned from `getAttribute` doesn't change ever if the user sets it explicitly. So if someone had an element that they wanted to make focusable but might dynamically set it to be disabled, hidden, or add/remove the `href`, `getAttribute("tabindex")` is always going to return `0`. I tested it out in the browser with `hidden` and missing `href`.  Because of that, I think this check needs to be moved to the bottom of the file because I could definitely see a use case where someone needs to make something focusable by default, but it's possible that it might become unfocusable/disabled/hidden later.

**Review by emplums — CHANGES_REQUESTED** [review]
This all looks good now, just left one more change I think we need to address before merging

File: docs/content/focusTrap.mdx
Comment (T-Hugs) [reply docs/content/focusTrap.mdx:51]: Already fixed 😅

**Comment by emplums** [comment]
re: https://github.com/primer/components/pull/1114/files#r590876019

@T-Hugs does the focus trap behavior need something to be initially focused to work? If not, I think it would make sense to remove the initial focus setting from this behavior and let that be something that is implemented in it's own hook/behavior and called in `useOverlay` separately. That way if folks want to use `useOpenAndCloseFocus` on something that doesn't trap focus it they aren't tied together. In my Overlay PR I implemented  a JS only `initialFocus` behavior and a`useOpenAndCloseFocus` hook which calls `initialFocus` and returns focus to an element when it is removed from the DOM using the useEffect cleanup syntax, so it's ready to go.

**Comment by T-Hugs** [comment]
@emplums Yes, when the focus trap is activated it is required to initially focus something (unless something inside the focus trap is already focused). The `initialFocus` argument lets you specify that element; otherwise it will default to the first tabbable element in the container.

I added this to support the case of confirmation dialogs (and similar UX), which may have a close button or other focusable UI above the OK/Cancel buttons.

I'm more than happy to take a second pass at some of these behaviors once they're all implemented in their base forms to make sure that we're sharing code where possible. As we work in parallel on many of these, there will be small bits of functionality that get duplicated.

**Comment by T-Hugs** [comment]
Opened two bugs to follow-up on issues in this PR:

* #1121 
* #1122

**Comment by emplums** [comment]
>  Yes, when the focus trap is activated it is required to initially focus something (unless something inside the focus trap is already focused). The initialFocus argument lets you specify that element; otherwise it will default to the first tabbable element in the container.

Sorry, didn't mean to ask if it's required a11y-wise, I mean is it required in order for this function to work correctly? Could we remove it and let another hook handle initial focus without breaking anything in this hook? If so, I think we should remove it. I see the focus trap behavior + the initial focus / return focus stuff being two separate concerns

**Review by emplums — APPROVED** [review]
ship it!

**Pull request merged**
